pulsar-broker/src/main/java/org/apache/pulsar/compaction/CompactedTopicImpl.java

```diff
@@ -317,9 +317,13 @@ private static int comparePositionAndMessageId(PositionImpl p, MessageIdData m)
             .compare(p.getEntryId(), m.getEntryId()).result();
     }
 
-    public synchronized Optional<Position> getCompactionHorizon() {
+    public Optional<Position> getCompactionHorizon() {
         return Optional.ofNullable(this.compactionHorizon);
     }
+
+    public CompletableFuture<CompactedTopicContext> getCompactedTopicContextFuture() {
+        return compactedTopicContext;
```

Comment:
The `compactedTopicContext` can be null. It's a risk of NPE for the caller. Because the caller will call this method in chain mode.

Reply:
We can add a `@Nullable` annotation.